aesd-char-driver/main.c

```diff
@@ -0,0 +1,311 @@
+    }
+
+    // Send entry found at the offset with remaining or max count
+    bytesNotCopied = copy_to_user(buf, (foundEntry->buffptr + offsetRtrn), bytesAvailToRead);
+
+    if(bytesNotCopied != 0)
+    {
+        printk(KERN_ALERT "bytes not coppied!!!");
+        retval = -ERESTARTSYS;
+        mutex_unlock(&local_aesd_dev_ptr->aesd_mutex);
+        goto exit_early_read;
+    }
+
+    retval = bytesAvailToRead; // we read the available bytes
+    *f_pos += bytesAvailToRead; // update file position
+    mutex_unlock(&local_aesd_dev_ptr->aesd_mutex); // unlock mutex for next transaction
+
+    exit_early_read:
+    return retval;
+}
+
+void freeCurrentInEntry(struct aesd_circular_buffer *buffer)
+{
+    //if(buffer->entry[buffer->in_offs] != NULL)
+    //{
+    printk(KERN_ALERT "freeing an entry");
+
+    kfree(buffer->entry[buffer->in_offs].buffptr);
+    buffer->entry[buffer->in_offs].size = 0;
+    buffer->entry[buffer->in_offs].buffptr = NULL;
+    //}
+}
+
+ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    ssize_t retval = -ENOMEM;
+    struct aesd_dev* local_aesd_dev_ptr;
+    ssize_t uncopBytes;
+    int i;
+    uint8_t newLineFound;
+
+
+    PDEBUG("write %zu bytes with offset %lld",count,*f_pos);
+    /**
+     * TODO: handle write
+     */
+    if(filp == NULL || buf == NULL || f_pos == NULL)
+    {
+        PDEBUG("Null pointer passed to write...");
+        retval = -EFAULT; 
+        goto exit_early_write;
+    }
+
+    local_aesd_dev_ptr = (struct aesd_dev*)(filp->private_data);
+
+
+    if(mutex_lock_interruptible(&(local_aesd_dev_ptr->aesd_mutex)))
+    {
+        printk(KERN_ALERT "Mutex for local dev filp private data didn't lock!!!");
+        retval = -ERESTARTSYS; 
+        goto exit_early_write;
+    }
+
+    
+    if(local_aesd_dev_ptr->entryToWrite.size == 0) // if nothing is being written
+    {
+        // kmalloc the buffer of count bytes
+        local_aesd_dev_ptr->entryToWrite.buffptr = kmalloc(count * sizeof(char), GFP_KERNEL);
+    }
+    else
+    {
+        // realloc to the new size count bytes
+        local_aesd_dev_ptr->entryToWrite.buffptr = krealloc(local_aesd_dev_ptr->entryToWrite.buffptr, (count + local_aesd_dev_ptr->entryToWrite.size)*sizeof(char), GFP_KERNEL);
+    }
+
+    if(local_aesd_dev_ptr->entryToWrite.buffptr == NULL)
+    {
+        retval = -ENOMEM;  // not enough memory
+        mutex_unlock(&local_aesd_dev_ptr->aesd_mutex); // unlock mutex for next transaction
+        goto exit_early_write;
+    }
+
+    // append to currently being written buffer
+    uncopBytes = copy_from_user((void*)(local_aesd_dev_ptr->entryToWrite.buffptr + local_aesd_dev_ptr->entryToWrite.size), buf, count);
+    local_aesd_dev_ptr->entryToWrite.size += count - uncopBytes; 
+
+    i = 0;
+    newLineFound = 0;
+    while (!newLineFound  && i < local_aesd_dev_ptr->entryToWrite.size)
```

Comment:
@Nicholas-Buckley77 with this logic I think you are going to return -ENOMEM when the `buf` passed does not contain an `\n`, which I think explains the errors output in the automated test runner log.